### PR TITLE
Improving the performance of reads from S3 by parallelizing downloads

### PR DESCRIPTION
**Pull request by rahul003**
- Updates the S3 File system in TensorFlow to read a file from S3 in parallel using the AWS SDK's transfer manager. 
- To support this PR, I made a PR to AWS SDK https://github.com/aws/aws-sdk-cpp/pull/1349, and have updated the version of SDK to match the release after that PR. 
- This PR also adds an override of the buffer size for TFRecordDataset similar to what's done for TPU GCS file system. This helps reduce overhead of network calls and thus improves performance, by reading larger parts of a file at once. 

**Benchmarking**
The following code snippet when run with TFRecords prepared from the ImageNet dataset with images resized to 480px (prepared using scripts [here](https://github.com/aws-samples/deep-learning-models/tree/master/utils/tensorflow)) gives **~2600 images/sec** when data is on S3. Same script when run with vanilla TF gives **~200 images/sec**, i.e. 13x speedup.

```
import tensorflow as tf
import pathlib
from timeit import default_timer as timer
import time
import os
import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--buffer-size', type=int, default=None)
parser.add_argument('--path-type', type=str, default='s3')
parser.add_argument('--path', type=str, default='s3://aws-tensorflow-benchmarking/imagenet-armand/train-480px/')
parser.add_argument('--num-batches', type=int, default=1000)
parser.add_argument('--batch-size', type=int, default=128)
parser.add_argument('--cache-processing', type=str, default='true')
args = parser.parse_args()

if args.path_type == 'local':
    data_dir = pathlib.Path('/home/ubuntu/imagenet-480px/')
    train_tdf = tf.data.TFRecordDataset([i.as_posix() for i in data_dir.glob('train*')], buffer_size=args.buffer_size)
elif args.path_type == 's3':
    files = tf.data.Dataset.list_files(os.path.join(args.path, 'train*'))
    train_tdf = tf.data.TFRecordDataset(files, buffer_size=args.buffer_size)
else:
    raise NotImplementedError()

features = {'image/encoded' : tf.io.FixedLenFeature((), tf.string, ""),
            'image/class/label': tf.io.FixedLenFeature([1], tf.int64,  -1),
                                         }
CACHE = {'image': None, 'label': None}

def parse(record):
    record = tf.io.parse_single_example(record, features)
    if args.cache_processing != 'true' or CACHE['image'] is None:
        image = record['image/encoded']
        image = tf.image.decode_jpeg(image)
        image = tf.image.resize(image, (224, 224))/255.
        CACHE['image'] = image
    else:
        image = CACHE['image']
    if args.cache_processing != 'true' or CACHE['label'] is None:
        label = record['image/class/label']
        label -= 1
        CACHE['label'] = label
    else:
        label = CACHE['label']
    return image, label


def benchmark(iterator):
    start_time = time.perf_counter()
    i = 0
    for sample in iterator:
        # Performing a training step
        i += 1
        if i% 10 == 0:
            print(f"Batch {i}: Time from start {time.perf_counter() - start_time}, Speed (imgs/sec): {i * args.batch_size / (time.perf_counter() - start_time)}")
        if i == args.num_batches:
            break
    cur_time = time.perf_counter()
    print(f"Total execution time: {cur_time - start_time}, Speed (imgs/sec):{i *args.batch_size / (cur_time - start_time)}")


iterator = train_tdf.map(parse, num_parallel_calls=tf.data.experimental.AUTOTUNE).batch(args.batch_size) \
                 .prefetch(tf.data.experimental.AUTOTUNE)

benchmark(iterator)
```

### CODE REVIEW

File: tensorflow/core/kernels/data/tf_record_dataset_op.cc [hunk 1]
```diff
@@ -255,7 +259,7 @@ void TFRecordDatasetOp::MakeDataset(OpKernelContext* ctx,
   OP_REQUIRES(ctx, buffer_size >= 0,
               errors::InvalidArgument(
                   "`buffer_size` must be >= 0 (0 == no buffering)"));
-
+
```
Comment (mihaimaruseac) [inline tensorflow/core/kernels/data/tf_record_dataset_op.cc:262]: Please remove trailing whitespace

File: third_party/aws/workspace.bzl [hunk 2]
```diff
@@ -9,11 +9,11 @@ def repo():
...
-            "https://mirror.bazel.build/github.com/aws/aws-sdk-cpp/archive/1.7.266.tar.gz",
-            "https://github.com/aws/aws-sdk-cpp/archive/1.7.266.tar.gz",
-        ],
-        sha256 = "39fd8a2999260d2b8fcbc8187f1ed5299972c2b8bd14adb7850fd674fea67fb7",
-        strip_prefix = "aws-sdk-cpp-1.7.266",
+             "https://mirror.bazel.build/github.com/aws/aws-sdk-cpp/archive/1.7.336.tar.gz",
+             "https://github.com/aws/aws-sdk-cpp/archive/1.7.336.tar.gz",
+         ],
+        sha256 = "758174f9788fed6cc1e266bcecb20bf738bd5ef1c3d646131c9ed15c2d6c5720",
+        strip_prefix = "aws-sdk-cpp-1.7.336",
```
Comment (mihaimaruseac) [inline third_party/aws/workspace.bzl:16]: Let's separate this into a different PR since it's an orthogonal change.

Comment (rahul003) [reply third_party/aws/workspace.bzl:16]: Actually the rest of the PR depends on this new version as I added a change in the SDK to enable download of a given range of file instead of the whole file only.

Comment (mihaimaruseac) [reply third_party/aws/workspace.bzl:16]: Makes sense.

Was thinking that we can add the SDK change first and then the rest of the PR. If the SDK causes internal failures we can roll that back easier.

But it's ok to keep as it is then.

**Review by mihaimaruseac — APPROVED** [review]
